<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN"
       "http://www.w3.org/TR/html4/loose.dtd">
<html>
<head>
<title>fastq-dump help</title>
</head>
<body>

<pre>
Usage:
  fastq-dump [options] [ -A ] &lt;accession>
  fastq-dump [options] &lt;path>

INPUT
  -A|--accession &lt;accession>       Replaces accession derived from &lt;path> in filename(s) and deflines
  --table &lt;table-name>             (New) Table name within SRA format, default is SEQUENCE

PROCESSING

Read Splitting                     Sequence data may be used as raw or split into individual reads
  --split-spot                     Split spots into individual reads

Full Spot Filters                  Applied to the full spot independently of --split-spot
  -N|--minSpotId &lt;rowid>           Minimum spot id
  -X|--maxSpotId &lt;rowid>           Maximum spot id
  --spot-groups &lt;[list]>           Filter by SPOT_GROUP (member): name[,...]
  -W|--clip                        Apply left and right clips

Common Filters                     Applied to spots when --split-spot is not set, otherwise - to individual reads
  -M|--minReadLen &lt;len>            Filter by sequence length >= &lt;len>
  -R|--read-filter &lt;[filter]>      Split into files by READ_FILTER value
                                   optionally filter by a value: pass|reject|criteria|redacted
  -E|--qual-filter                 Filter used in early 1K Genomes data:
                                   no sequences starting or ending with >= 10N

Filters for individual reads       Applied only with --split-spot set
  --skip-technical                 Dump only biological reads

OUTPUT
  -O|--outdir &lt;path>               Output directory, default is '.'

Multiple File Options              Setting this options will produce more than 1 file, which will be suffixed by splitting criteria.
  --split-files                    Dump each read into a separate file.Files will received suffix corresponding to read number
  --split-3                        Legacy 3-file splitting for mate-pairs:
                                   First 2 biological reads satisfying dumping conditions
                                   are placed in files *_1.fastq and *_2.fastq
                                   If only 1 biological read is dumpable - it is placed in *.fastq
                                   Biological reads 3 and above are ignored.
  -G|--spot-group                  Split into files by SPOT_GROUP (member name)
  -R|--read-filter &lt;[filter]>      Split into files by READ_FILTER value
                                   optionally filter by a value: pass|reject|criteria|redacted
  -T|--group-in-dirs               Split into subdirectories instead of files
  -K|--keep-empty-files            Do not delete empty files

FORMATTING

Sequence
  -C|--dumpcs &lt;[cskey]>            Formats sequence in color space (default for SOLiD),cskey may be specified for translation
  -B|--dumpbase                    Formats sequence in base sequence (default for other than SOLiD).

Quality
  -Q|--offset &lt;integer>            Offset to use for quality conversion, default is 33
  --fasta                          Fasta only, no qualities

Defline
  -F|--origfmt                     Defline contains only original sequence name
  -I|--readids                     Append read id after spot id as 'accession.spot.readid' on defline
  --helicos                        Helicos style defline
  --defline-seq &lt;fmt>              Defline format specification for sequence.
  --defline-qual &lt;fmt>             Defline format specification for quailty.
                                   &lt;fmt> is string of characters and/or variables. Variables could be are one of:
                                      $ac - accession, $si - spot id, $sn - spot name, $sg - spot group (barcode),
                                      $sl - spot length in bases, $ri - read number, $rn - read name, $rl - read length in bases.
                                   '[]' could be used for an optional output: if all vars in [] yield empty values whole group is not printed.
                                   Empty value is empty string or 0 for numeric variables.
                                   Ex: @$sn[_$rn]/$ri - '_$rn' is omitted if name is empty

OTHER:
  -h|--help                        Output a brief explantion for the program
  -V|--version                     Display the version of the program then quit
  -L|--log-level &lt;level>           Logging level as number or enum string
                                   One of (fatal|sys|int|err|warn|info) or (0-5)
                                   Current/default is warn
  -v|--verbose                     Increase the verbosity level of the program
                                   Use multiple times for more verbosity
</pre>
</body>
</html>
